rfcs/text/0010_service_status.md

```diff
@@ -0,0 +1,304 @@
+- Start Date: 2020-03-07
+- RFC PR: https://github.com/elastic/kibana/pull/59621
+- Kibana Issue: https://github.com/elastic/kibana/issues/41983
+
+# Summary
+
+A set API for describing the current status of a system (Core service or plugin)
+in Kibana.
+
+# Basic example
+
+```ts
+// Override default behavior and only elevate severity when elasticsearch is not available
+core.status.set(
+  core.status.core$.pipe(core => core.elasticsearch);
+)
+```
+
+# Motivation
+
+Kibana should do as much possible to help users keep their installation in a working state. This includes providing as much detail about components that are not working as well as ensuring that failures in one part of the application do not block using other portions of the application.
+
+In order to provide the user with as much detail as possible about any systems that are not working correctly, the status mechanism should provide excellent defaults in terms of expressing relationships between services and presenting detailed information to the user.
+
+# Detailed design
+
+## Types
+
+```ts
+/**
+ * The current status of a service at a point in time.
+ */
+type ServiceStatus = {
+  /** The current availability level of the service. */
+  level: ServiceStatusLevel.available;
+  /** A high-level summary of the service status. */
+  summary?: string;
+  /** A more detailed description of the service status. */
+  detail?: string;
+  /** A URL to open in a new tab about how to resolve or troubleshoot the problem */
+  documentationUrl?: string;
+  /** Any JSON-serializable data to be included in the HTTP API response. */
+  meta?: object;
+} | {
+  /** The current availability level of the service. */
+  level: ServiceStatusLevel;
+  /** A high-level summary of the service status. */
+  summary: string; // required when level !== available
+  /** A more detailed description of the service status. */
+  detail?: string;
+  /** A URL to open in a new tab about how to resolve or troubleshoot the problem */
+  documentationUrl?: string;
+  /** Any JSON-serializable data to be included in the HTTP API response. */
+  meta?: object;
+}
+
+/**
+ * The current "level" of availability of a service.
+ */
+enum ServiceStatusLevel {
+  available,   // everything is working!
+  degraded,    // some features may not be working
+  unavailable, // the service is unavailable, but other functions should work
+  fatal        // block all user functions and display the status page
```

Comment:
is this recoverable error? or a user has to reload the Kibana server?

Reply:
It could be recoverable. `fatal` is probably not the right word. What about `blocking` or `critical`?

Reply:
Changed to `critical`, but let me know if you have a better suggestion.